snap/validate_test.go

```diff
@@ -692,6 +695,91 @@ layout:
 		c.Assert(err, ErrorMatches, `layout "/usr/foo/bar" underneath prior layout item "/usr/foo"`)
 	}
 
+	const yaml6 = `
+name: tricky-layout-1
+layout:
+  /etc/norf:
+    bind: $SNAP/etc/norf
+  /etc/norf:
+    bind-file: $SNAP/etc/norf
+`
+	info, err := InfoFromSnapYamlWithSideInfo([]byte(yaml6), &SideInfo{Revision: R(42)})
+	c.Assert(err, IsNil)
+	err = ValidateLayoutAll(info)
+	c.Assert(err, IsNil)
+	c.Assert(info.Layout, HasLen, 1)
+	c.Assert(info.Layout["/etc/norf"].Bind, Equals, "")
+	c.Assert(info.Layout["/etc/norf"].BindFile, Equals, "$SNAP/etc/norf")
+
+	// Two layouts refer to the same path as a directory and a file.
+	const yaml7 = `
+name: clashing-source-path-1
+layout:
+  /etc/norf:
+    bind: $SNAP/etc/norf
+  /etc/corge:
+    bind-file: $SNAP/etc/norf
+`
+	info, err = InfoFromSnapYamlWithSideInfo([]byte(yaml7), &SideInfo{Revision: R(42)})
+	c.Assert(err, IsNil)
+	err = ValidateLayoutAll(info)
+	c.Assert(err, ErrorMatches, `layout "/etc/norf" refers to directory "\$SNAP/etc/norf" but another layout treats it as file`)
+
+	// Two layouts refer to the same path as a directory and a file (other way around).
+	const yaml8 = `
+name: clashing-source-path-2
+layout:
+  /etc/norf:
+    bind-file: $SNAP/etc/norf
+  /etc/corge:
+    bind: $SNAP/etc/norf
+`
+	info, err = InfoFromSnapYamlWithSideInfo([]byte(yaml8), &SideInfo{Revision: R(42)})
+	c.Assert(err, IsNil)
+	err = ValidateLayoutAll(info)
+	c.Assert(err, ErrorMatches, `layout "/etc/norf" refers to file "\$SNAP/etc/norf" but another layout treats it as a directory`)
+
+	// Two layouts refer to the same path, but one uses variable and the other doesn't.
```

Comment:
Nice

Reply:
I was happy I could write this test and it just worked :)